orchagent/pfcwdorch.cpp

```diff
@@ -0,0 +1,755 @@
+            else
+            {
+                normalQueues.push_back(sai_serialize_object_id(queueId));
+            }
+        }
+    }
+
+    // Run scripts for selected queues to see if their state changed
+    // from normal to stormed and vice versa
+    set<string> stormCheckReply;
+    set<string> restoreCheckReply;
+    vector<string> argv =
+    {
+        to_string(COUNTERS_DB),
+        COUNTERS_TABLE,
+        to_string(nearestTime * 1000)
+    };
+
+    if (!normalQueues.empty())
+    {
+        stormCheckReply = runRedisScript(db, detectSha, normalQueues, argv);
+    }
+
+    if (!stormedQueues.empty())
+    {
+        restoreCheckReply = runRedisScript(db, restoreSha, stormedQueues, argv);
+    }
+
+    // Update internal state of queues and their time
+    for (auto& queueKv : m_entryMap)
+    {
+        sai_object_id_t queueId = queueKv.first;
+        PfcWdQueueEntry& queueEntry = queueKv.second;
+
+        string queueIdStr = sai_serialize_object_id(queueId);
+
+        // Queue became stormed
+        if (stormCheckReply.find(queueIdStr) != stormCheckReply.end())
+        {
+            if (queueEntry.c_action == PfcWdAction::PFC_WD_ACTION_DROP)
+            {
+                queueEntry.handler = make_shared<DropHandler>(
+                        queueEntry.portId,
+                        queueId,
+                        queueEntry.index,
+                        PfcWdOrch<DropHandler, ForwardHandler>::getCountersTable());
+            }
+            else if (queueEntry.c_action == PfcWdAction::PFC_WD_ACTION_FORWARD)
+            {
+                queueEntry.handler = make_shared<ForwardHandler>(
+                        queueEntry.portId,
+                        queueId,
+                        queueEntry.index,
+                        PfcWdOrch<DropHandler, ForwardHandler>::getCountersTable());
+            }
+            else
+            {
+                throw runtime_error("Invalid PFC WD Action");
+            }
+
+            queueEntry.pollTimeLeft = queueEntry.c_restorationTime;
+        }
+        // Queue is restored
+        else if (restoreCheckReply.find(queueIdStr) != restoreCheckReply.end())
+        {
+            queueEntry.handler = nullptr;
+            queueEntry.pollTimeLeft = queueEntry.c_detectionTime;
+        }
+        // Update queue poll timer
+        else
+        {
+            queueEntry.pollTimeLeft = queueEntry.pollTimeLeft == nearestTime ?
+                (queueEntry.handler == nullptr ?
+                 queueEntry.c_detectionTime :
+                 queueEntry.c_restorationTime) :
+                queueEntry.pollTimeLeft - nearestTime;
+        }
+    }
+}
+
+template <typename DropHandler, typename ForwardHandler>
+void PfcWdSwOrch<DropHandler, ForwardHandler>::pfcWatchdogThread(void)
+{
+    DBConnector db(COUNTERS_DB, DBConnector::DEFAULT_UNIXSOCKET, 0);
+
+    // Load script for storm detection
+    string detectScriptName = getStormDetectionCriteria();
```

Comment:
how can I implement a different detection script?

Reply:
Inherit from base class and specify other script. However, scripts will be moved to syncd in next iteration and their names will be derived from asic.